### PR TITLE
adding more commands to clarify steps / stages

### PR DESCRIPTION
**Pull request by jluk**
**What this PR does / why we need it**:
1. There is some missing pre-reqs, like configuring a key vault. Added a section/doc link.
1. Important data fields from Azure should be enumerated earlier in the doc, as its needed for swapping yaml out.
1. Added SP creation steps & env vars for setup on SP option

**Special notes for your reviewer**:
I'm did not get it working, as my pod is stuck in deployment - so do gut check me on the proposed changes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,26 +166,41 @@ The Azure Key Vault Provider offers four modes for accessing a Key Vault instanc
 
 > Supported with linux and windows
 
-1. Add your service principal credentials as a Kubernetes secrets accessible by the Secrets Store CSI driver.
+1. Add your service principal credentials as a Kubernetes secrets accessible by the Secrets Store CSI driver. If using AKS you can learn about [service principals in AKS here.](https://docs.microsoft.com/azure/aks/kubernetes-service-principal)
```
Comment (ritazh) [inline README.md:169]: This is great! Can you also add https://docs.microsoft.com/en-us/azure/aks/use-managed-identity for AKS for the other two options: system assigned and user assigned identity.

File: README.md [hunk 2]
```diff
@@ -114,21 +129,35 @@ Create a `secretproviderclasses` resource to provide provider-specific parameter
...
-    ```yaml
-    volumes:
-      - name: secrets-store-inline
-        csi:
-          driver: secrets-store.csi.k8s.io
-          readOnly: true
-          volumeAttributes:
-            secretProviderClass: "azure-kvname"
-    ```
+    If you did not change the name of the secretProviderClass previously, no changes are needed.
```
Comment (aramase) [inline README.md:134]: nit: This line needs to be right indented by one tab to preserve the yaml formatting below.

File: README.md [hunk 3]
```diff
@@ -137,26 +166,41 @@ The Azure Key Vault Provider offers four modes for accessing a Key Vault instanc
...
     Ensure this service principal has all the required permissions to access content in your Azure key vault instance.
-    If not, you can run the following using the Azure cli:
+    If you do not have a service principal, you can run the following Azure CLI commands to create a new service principal and assign permissions:
 
     ```bash
-    # Assign Reader Role to the service principal for your keyvault
-    az role assignment create --role Reader --assignee <principalid> --scope /subscriptions/<subscriptionid>/resourcegroups/<resourcegroup>/providers/Microsoft.KeyVault/vaults/<keyvaultname>
+    # Set environment variables
+    SPNAME=<servicePrincipalName>
+    APPID=<service-principal-appId>
```
Comment (aramase) [inline README.md:184]: can we replace this with 
```
APPID=$(az ad sp show --id http://${SPNAME} --query appId -otsv)
```

File: README.md (hunk 1 above)
Comment (jluk) [reply README.md:169]: good idea, adding for system assigned and adding a note AKS has a gap to support UMI.

File: README.md (hunk 2 above)
Comment (jluk) [reply README.md:134]: ack

File: README.md (hunk 3 above)
Comment (jluk) [reply README.md:184]: good idea, added and moved the sp creation step to first step

**Comment by jluk** [comment]
> > I'm did not get it working, as my pod is stuck in deployment - so do gut check me on the proposed changes.
> 
> @jluk Are there specific issues with the deployment that we can help you troubleshoot?

My deployment of `nginx-secrets-store-inline` failed due to a mount failure. It seems the default config to have "cloudname" is causing an issue?

```
  cloudName: "cloudname"          # [OPTIONAL for Azure] if not provided, azure environment will default to AzurePublic 
```

Pod describe for nginx stuck in creating state:
```
Events:
  Type     Reason       Age                    From                                        Message
  ----     ------       ----                   ----                                        -------
  Warning  FailedMount  14m (x285 over 18h)    kubelet, aks-nodepool1-93144144-vmss000005  Unable to attach or mount volumes: unmounted volumes=[secrets-store-inline], unattached volumes=[secrets-store-inline default-token-m7ct6]: timed out waiting for the condition
  Warning  FailedMount  5m24s (x699 over 18h)  kubelet, aks-nodepool1-93144144-vmss000005  (combined from similar events): MountVolume.SetUp failed for volume "secrets-store-inline" : kubernetes.io/csi: mounter.SetupAt failed: rpc error: code = Unknown desc = error mounting secret time="2020-05-06T16:40:14Z" level=fatal msg="[error] : cloudName cloudname is not valid, error: autorest/azure: There is no cloud environment matching the name \"CLOUDNAME\""
 for pod: 6edc0cb0-f090-4b01-8688-82439e6a9f4f, ns: default
```

**Comment by aramase** [comment]
> EDIT: I realize now the cloudname is just a placeholder, perhaps we should just leave it blank in the example so it defaults?

That's right. If left blank will default to `AzurePublicCloud`. Agree, we should keep it blank in docs to avoid confusion.

**Comment by jluk** [comment]
@aramase 

* The deployment worked if I leave it an empty string, but I actually tried to place "AzurePublic" according to this comment and it failed with similar error as above claiming it needs to match "AZUREPUBLIC". 
```
Events:
  Type     Reason       Age                   From                                        Message
  ----     ------       ----                  ----                                        -------
  Warning  FailedMount  31m (x3 over 44m)     kubelet, aks-nodepool1-93144144-vmss000005  Unable to attach or mount volumes: unmounted volumes=[secrets-store-inline], unattached volumes=[default-token-m7ct6 secrets-store-inline]: timed out waiting for the condition
  Warning  FailedMount  6m37s (x35 over 65m)  kubelet, aks-nodepool1-93144144-vmss000005  (combined from similar events): MountVolume.SetUp failed for volume "secrets-store-inline" : kubernetes.io/csi: mounter.SetupAt failed: rpc error: code = Unknown desc = error mounting secret time="2020-05-06T17:58:04Z" level=fatal msg="[error] : cloudName AzurePublic is not valid, error: autorest/azure: There is no cloud environment matching the name \"AZUREPUBLIC\""
 for pod: ad12aca8-8a78-4fcd-a1c5-446dbdb5d540, ns: default
  Warning  FailedMount  112s (x21 over 69m)  kubelet, aks-nodepool1-93144144-vmss000005  Unable to attach or mount volumes: unmounted volumes=[secrets-store-inline], unattached volumes=[secrets-store-inline default-token-m7ct6]: timed out waiting for the condition
```

Is that expected and if so should we update the comments?
```
    cloudName: ""          # [OPTIONAL for Azure] if not provided, azure environment will default to AzurePublic 
```

* Is there a good exercise a user can do after deployment to validate everything is working? I'm not clear how to access the keys needed and prove that it's working beyond showing all deployments are successful and a describe on the secretproviderclass.

**Comment by aramase** [comment]
@jluk 

> The deployment worked if I leave it an empty string, but I actually tried to place "AzurePublic" according to this comment and it failed with similar error as above claiming it needs to match "AZUREPUBLIC".

The allowed values are documented in the table - 
```
[available for version > 0.0.4] name of the azure cloud based on azure go sdk (AzurePublicCloud,AzureUSGovernmentCloud, AzureChinaCloud, AzureGermanCloud)
```

It needs to contain the `Cloud` suffix. 

> Is there a good exercise a user can do after deployment to validate everything is working? I'm not clear how to access the keys needed and prove that it's working beyond showing all deployments are successful and a describe on the secretproviderclass.

A good way to validate it would be by running -  

```
kubectl exec -it nginx-secrets-store-inline ls /mnt/secrets-store/
secret1

kubectl exec -it nginx-secrets-store-inline cat /mnt/secrets-store/secret1
test
```

This verification step is documented for `pod-identity` mode. We should probably document it in a common place and refer to it in all the modes.

**Comment by jluk** [comment]
Thanks @aramase added the changes for everything mentioned LMK thoughts.

Assuming the expected output of the exec is:
```
➜  Desktop kubectl exec -it nginx-secrets-store-inline ls /mnt/secrets-store/test-secret
+ kubectl exec -it nginx-secrets-store-inline ls /mnt/secrets-store/test-secret
/mnt/secrets-store/test-secret
```

File: README.md [hunk 4]
```diff
@@ -114,21 +129,51 @@ Create a `secretproviderclasses` resource to provide provider-specific parameter
...
+        volumes:
+          - name: secrets-store-inline
+            csi:
+              driver: secrets-store.csi.k8s.io
+              readOnly: true
+              volumeAttributes:
+                secretProviderClass: "azure-kvname"
+        ```
+
+  1. Select and complete an option from [below to enable access to the Key Vault](####Provide-Identity-to-Access-Key-Vault)
```
Comment (aramase) [inline README.md:146]: nit: Needs to be `[below to enable access to the Key Vault](#provide-identity-to-access-key-vault)`

File: README.md [hunk 5]
```diff
@@ -114,21 +129,51 @@ Create a `secretproviderclasses` resource to provide provider-specific parameter
...
+
+    ```bash
+    kubectl exec -it nginx-secrets-store-inline-podid ls /mnt/secrets-store/
+    secret1
+    ```
+
+    You should see an output similar to:
+    ```bash
+    ➜  Desktop kubectl exec -it nginx-secrets-store-inline ls /mnt/secrets-store/secret1
+    + kubectl exec -it nginx-secrets-store-inline ls /mnt/secrets-store/secret1
```
Comment (aramase) [inline README.md:168]:
```
kubectl exec -it nginx-secrets-store-inline cat /mnt/secrets-store/secret1
test
```

The second snippet for displaying the secret value

File: README.md [hunk 6]
```diff
@@ -114,21 +129,51 @@ Create a `secretproviderclasses` resource to provide provider-specific parameter
...
+1. Validate the pod has access to the secret from key vault:
+
+    ```bash
+    kubectl exec -it nginx-secrets-store-inline-podid ls /mnt/secrets-store/
+    secret1
+    ```
+
+    You should see an output similar to:
+    ```bash
+    ➜  Desktop kubectl exec -it nginx-secrets-store-inline ls /mnt/secrets-store/secret1
```
Comment (aramase) [inline README.md:167]: nit: can we remove this?

File: README.md [hunk 7]
```diff
@@ -114,21 +129,51 @@ Create a `secretproviderclasses` resource to provide provider-specific parameter
...
+
+  1. Deploy the secretProviderClass
+
+     `kubectl apply -f ./examples/v1alpha1_secretproviderclass.yaml`
+
+  1. Deploy the Linux deployment yaml
+
+     `kubectl apply -f ./examples/nginx-pod-secrets-store-inline-volume-secretproviderclass.yaml`
+
+#### Validate the secret
```
Comment (ritazh) [inline README.md:156]: we can perhaps consider using the same stuff content from https://github.com/kubernetes-sigs/secrets-store-csi-driver#secret-content-is-mounted-on-pod-start

**Comment by jluk** [comment]
made some chgs over lunch lmk thoughts @ritazh / @aramase

File: README.md (hunk 7 above)
Comment (aramase) [reply README.md:156]: @jluk We should probably copy this part of the content from the link @ritazh posted -

To validate, once the pod is started, you should see the new mounted content at the volume path specified in your deployment yaml.
```
kubectl exec -it nginx-secrets-store-inline ls /mnt/secrets-store/
foo
```

I don't think we need a link to the driver readme (just so we can avoid the extra hop to different readme for reading about validation). WDYT?

Comment (jluk) [reply README.md:156]: made another change per ask

Comment (aramase) [reply README.md:156]: Thank you @jluk! Looks good to me.

**Pull request merged**